paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc

```diff
@@ -0,0 +1,163 @@
+  // ijkl,z(kl),z(kl)->ijkl (x,scale,bias->out, begin_norm_axis=2, z=kl)
+  std::string x_axes = "";
+  for (auto i = 0; i < x_ndim; ++i) {
+    x_axes += static_cast<char>(static_cast<int>('k') - begin_norm_axis + i);
+  }
+
+  std::string scale_axes;
+  std::string bias_axes;
+  if (x_ndim - begin_norm_axis == 1) {
+    scale_axes = "k";
+    bias_axes = "k";
+  } else {
+    // z = x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+    scale_axes = "z";
+    bias_axes = "z";
+  }
+  std::string out_axes = x_axes;
+
+  VLOG(4) << "LayerNormSPMDRule build Einsum notation (x,scale,bias->out): ["
+          << x_axes << "," << scale_axes << "," << bias_axes << " --> "
+          << out_axes << "](begin_norm_axis:" << begin_norm_axis
+          << ",z=" << x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+          << ").";
+
+  // step2: Sharding Propogation
+
+  TensorDistAttr output_dist_attr_dst =
+      CopyTensorDistAttrForOutput(x_dist_attr_src);
+  TensorDistAttr x_dist_attr_dst = CopyTensorDistAttrForOutput(x_dist_attr_src);
+  std::vector<int64_t> out_dims_mapping;
+  out_dims_mapping.reserve(out_axes.size());
+  for (size_t i = 0; i < out_axes.size(); ++i) {
+    if (i < static_cast<size_t>(begin_norm_axis)) {
+      out_dims_mapping.push_back(x_dims_mapping[i]);
+    } else {
+      out_dims_mapping.push_back(-1);
+    }
+  }
+  output_dist_attr_dst.set_dims_mapping(out_dims_mapping);
+  x_dist_attr_dst.set_dims_mapping(out_dims_mapping);
+
+  // step2.3: Merge and get Inputs' New Dims Mapping.
+  input_dist_attrs.emplace_back(x_dist_attr_dst);
+
+  // TODO(zhiqiu): support shardding on scale and bias
+  // Now, apply replicating.
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[1].dist_attr()));
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[2].dist_attr()));
+
+  // Step2.4.  handle input and out tensor partial
+  std::vector<int64_t> partial_on_dims;
+  for (size_t i = 0; i < out_dims_mapping.size(); ++i) {
+    if (out_dims_mapping[i] > -1) {
+      partial_on_dims.push_back(out_dims_mapping[i]);
+    }
+  }
+
+  VLOG(4) << "LayerNormSPMDRule InferForward: "
+          << "X shape: [" << str_join(x_shape) << "], src_dims_mapping: ["
+          << str_join(x_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(x_dist_attr_dst.dims_mapping()) << "]; scale shape: ["
+          << str_join(scale_shape) << "], src_dims_mapping: ["
+          << str_join(scale_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(input_dist_attrs[1].dims_mapping()) << "]; bias shape: ["
+          << str_join(bias_shape) << "], src_dims_mapping: ["
+          << str_join(bias_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(input_dist_attrs[2].dims_mapping())
+          << "]; out dims_mapping: [" << str_join(out_dims_mapping)
+          << "], partial_on_dims: [" << str_join(partial_on_dims) << "]";
+
+  return {input_dist_attrs, {output_dist_attr_dst}};
```

Comment:
should infer the distattr of output （variance and mean）

Reply:
done thx